sql/parser/builtins.go

```diff
@@ -1405,32 +1401,79 @@ func overlay(s, to string, pos, size int) (Datum, error) {
 }
 
 func round(x float64, n int64) (Datum, error) {
-	switch {
-	case n < 0:
-		return DNull, errRoundNumberDigits
-	case n > 323:
-		// When rounding to more than 323 digits after the decimal
-		// point, the original number is returned, because rounding has
-		// no effect at scales smaller than 1e-323.
-		//
-		// 323 is the sum of
-		//
-		// 15, the maximum number of significant digits in a decimal
-		// string that can be converted to the IEEE 754 double precision
-		// representation and back to a string that matches the
-		// original; and
-		//
-		// 308, the largest exponent. The significant digits can be
-		// right shifted by 308 positions at most, by setting the
-		// exponent to -308.
+	pow := math.Pow(10, float64(n))
+
+	if pow == 0 {
+		// Rounding to so many digits on the left that we're underflowing.
+		// Avoid a NaN below.
+		return NewDFloat(DFloat(0)), nil
+	}
+	if math.Abs(x*pow) > 1e17 {
+		// Rounding touches decimals below float precision; the operation
+		// is a no-op.
 		return NewDFloat(DFloat(x)), nil
 	}
-	const b = 64
-	y, err := strconv.ParseFloat(strconv.FormatFloat(x, 'f', int(n), b), b)
-	if err != nil {
-		panic(fmt.Sprintf("parsing a float that was just formatted failed: %s", err))
+
+	v, frac := math.Modf(x * pow)
+	if x > 0.0 {
+		if frac > 0.5 || (frac == 0.5 && uint64(v)%2 != 0) {
+			v += 1.0
+		}
+	} else {
+		if frac < -0.5 || (frac == -0.5 && uint64(v)%2 != 0) {
+			v -= 1.0
+		}
+	}
+
+	return NewDFloat(DFloat(v / pow)), nil
+}
+
+const (
+	scaleRatio = math.Ln2 / math.Ln10
+	// Compute the size of a Word in bits.
+	_m       = ^big.Word(0)
+	_logS    = _m>>8&1 + _m>>16&1 + _m>>32&1
+	wordSize = 8 * (1 << _logS)
```

Comment:
Any suggestion as to how to implement this without having `go vet` bark on `_m >> 32` like it currently does?